kdc/krb5tgs.c

```diff
@@ -807,7 +807,7 @@ tgs_make_reply(krb5_context context,
     if(ret)
 	goto out;
 
-    ret = copy_Realm(&server_principal->realm, &rep.ticket.realm);
+    ret = copy_Realm(&server->entry.principal->realm, &rep.ticket.realm);
```

Comment:
Hmm, is this just needed to handle referrals? If I remember right, as of W2K3 the canonicalize flag is ignored for TGS service ticket requests, unless you're returning a referral. How about this?

```
diff --git a/kdc/krb5tgs.c b/kdc/krb5tgs.c
index 51f0c9208..1fcdd5e6f 100644
--- a/kdc/krb5tgs.c
+++ b/kdc/krb5tgs.c
@@ -755,7 +755,7 @@ tgs_make_reply(krb5_context context,
     EncTicketPart et;
     KDCOptions f = b->kdc_options;
     krb5_error_code ret;
-    int is_weak = 0;
+    int is_weak = 0, tmp;
 
     memset(&rep, 0, sizeof(rep));
     memset(&et, 0, sizeof(et));
@@ -807,7 +807,11 @@ tgs_make_reply(krb5_context context,
     if(ret)
        goto out;
 
-    ret = copy_Realm(&server_principal->realm, &rep.ticket.realm);
+    if (krb5_find_padata(enc_pa_data->val, enc_pa_data->len,
+                        KRB5_PADATA_SERVER_REFERRAL, &tmp))
+       ret = copy_Realm(&server->entry.principal->realm, &rep.ticket.realm);
+    else
+       ret = copy_Realm(&server_principal->realm, &rep.ticket.realm);
     if (ret)
        goto out;
     _krb5_principal2principalname(&rep.ticket.sname, server_principal);
@@ -1108,7 +1112,8 @@ need_referral(krb5_context context, krb5_kdc_configuration *config,
 {
     const char *name;
 
-    if(!options->canonicalize && server->name.name_type != KRB5_NT_SRV_INST)
+    if(!options->canonicalize && !options->enc_tkt_in_skey &&
+       server->name.name_type != KRB5_NT_SRV_INST)
        return FALSE;
 
     if (server->name.name_string.len == 1)
```

Checking `enc_tkt_in_skey` is unset mirrors the MIT code. If this breaks any of your tests, let me know more detail.

Reply:
So, yeah, in principle the `Ticket`'s `sname` and `realm` can be anything at all because the `Ticket` is opaque.  And it's wrong for _clients_ to check if the `KDC-REP`'s `Ticket`'s `sname` and `realm` matches the one requested.  So I'm OK with this change.

@lhoward PA-SVR-REFERRAL-DATA is old.  Clients aren't supposed to use it.

Or, actually, what's this change _really_ about?  I'm kinda tired, so I'm probably missing something.

Reply:
PA-SVR-REFERRAL-DATA check was just to see whether referrals were being returned, without adding yet another argument to tgs_make_reply(). could add a flag instead.

Reply:
Ah, ok.  But do you agree that it almost doesn't matter?  Or do we need to worry about clients that do inspect the issued `Ticket`?

Reply:
@iboukris / @abartlet does Windows always canonicalize the realm on a TGS-REQ even when the canonicalize flag is unset?

Reply:
As far as I can tell, realm canonicalization such of netbios and lowercase-realm is independent of the canonicalize flag, and is *always* done by Windows.

The current case is when a ticket for netbios name is requested, which can be done with:
$ ./kuser/kgetcred HTTP/apache.acme.com@ACME

The capath function would think it's a foreign realm and get a cross realm ticket first, but at the end the KDC returned ticket realm is the canonical one ACME.COM.

Reply:
I recall that Windows always canonicalizes the realm in an AS-REQ, but I wasn't sure about TGS-REQ in the non-referral case. Can you try ./kuser/kgetcred HTTP/apache.acme.com@acme.com and see what you get back?

Reply:
And also what is actually in the server ticket if possible.

Reply:
Assuming you have authenticated as a principal in ACME.COM that is.

Reply:
OK, I guess this is fine then. MIT and Heimdal have different philosophies on whether name canonicalization should be done by the front- or back-end and I suppose there is not a right answer. In the AS-REQ path MIT will return the requested client name if canonicalize is not requested, but is has a special case for canonicalizing the realm for dealing with Windows; in the TGS-REQ path, it only canonicalizes when issuing referrals. Heimdal always returns the canonical principal, that is, the principal the backend returned, and it's up to the backend to *not* do that if (for example) HDB_F_CANON is not requested.

@nicowilliams do you have any further thoughts? I guess we at least should be consistent between AS and TGS paths, which this patch does. Myself, I don't really mind putting Windows-specific stuff in the KDC front-end as there is already so much anyway, but perhaps there's an argument that it's nice to avoid.

Reply:
I am fine with this change as-is.

Reply:
I have no further thoughts as to this patch.  I do think we should implement AD's case-insensitivity as to realms.  Case-insensitivity for other things (such as hostname and domainname components of host-based and domain-based service names) might be nice, but much harder to implement.